### PR TITLE
Clean up based on the String Processing Algorithms proposal

### PR DESCRIPTION
**Pull request by itingliu**
- Move functions and types that have not been proposed from public to internal
- Add doc comments for public API
- Add FIXME for API awaiting SE-0346
- Replace `_MatchResult` with `Regex<Output>.Match` and update tests

### CODE REVIEW

File: Sources/_StringProcessing/Algorithms/Matching/MatchReplace.swift [hunk 1]
```diff
@@ -110,4 +110,82 @@ extension RangeReplaceableCollection where SubSequence == Substring {
...
+  ///   - replacement: A closure that receives the full match information,
+  ///   including captures, and returns a replacement collection.
+  ///   - subrange: The range in the collection in which to search for `regex`.
+  ///   - maxReplacements: A number specifying how many occurrences of the
+  ///   sequence matching `regex` to replace. Default is `Int.max`.
+  /// - Returns: A new collection in which all occurrences of subsequence
+  /// matching `regex` are replaced by `replacement`.
+  public func replacing<R: RegexComponent, Replacement: Collection>(
+    _ regex: R,
+    with replacement: (Regex<R.Output>.Match) throws -> Replacement,
```
Comment (itingliu) [inline Sources/_StringProcessing/Algorithms/Matching/MatchReplace.swift:127]: @milseman this would replace `_MatchResult` boilerplate

**Review by rxwei — APPROVED** [review]

**Comment by itingliu** [comment]
@swift-ci please test

File: Sources/_StringProcessing/Algorithms/Matching/MatchReplace.swift (hunk 1 above)
Comment (milseman) [reply Sources/_StringProcessing/Algorithms/Matching/MatchReplace.swift:127]: That's great!

File: Tests/RegexBuilderTests/AlgorithmsTests.swift [hunk 2]
```diff
@@ -38,13 +40,13 @@ class RegexConsumerTests: XCTestCase {
       int,
       input: "foo 160 bar 99 baz",
       result: "foo 240 bar 143 baz",
-      { match in String(match.result.1, radix: 8) })
+      { match in String(match.output.1, radix: 8) })
```
Comment (milseman) [inline Tests/RegexBuilderTests/AlgorithmsTests.swift:43]: Just to make sure, couldn't this be `match.1`?, etc? No need to synthesize the entire output.

**Review by milseman — APPROVED** [review]
Overall LGTM, thanks for internalizing all of this.

Future work would be to see what we should remove, what we should underscore, etc., but getting this out of the module's API is a huge step. Thanks!

File: Tests/RegexBuilderTests/AlgorithmsTests.swift (hunk 2 above)
Comment (itingliu) [reply Tests/RegexBuilderTests/AlgorithmsTests.swift:43]: Yeah you're right that this can just be `match.1`. I was just doing a blind replace.

**Comment by itingliu** [comment]
@swift-ci please test

**Comment by itingliu** [comment]
@swift-ci Please test

**Comment by hamishknight** [comment]
@swift-ci please test

**Pull request merged**